lib/ddtrace/chunker.rb

```diff
@@ -0,0 +1,37 @@
+require 'json'
+require 'msgpack'
+
+module Datadog
+  # Chunks list of elements into batches
+  module Chunker
+    module_function
+
+    # Chunks a list into batches of at most +max_chunk_size+ elements each.
+    #
+    # An exception can occur if a single element is too large. That single
+    # element will be returned in its own chunk. You have to verify by yourself
+    # when such elements are returned.
+    #
+    # @param list [Enumerable] list of elements
+    # @param max_chunk_size [Numeric] maximum acceptable chunk size
+    # @return [Enumerator] lazy list of chunks
+    def chunk_by_size(list, max_chunk_size)
+      chunk_agg = 0
+      list.slice_before do |elem|
+        size = elem.size
```

Comment:
There's an assumption that every item in the list responds to `size`; is this too specific?

Reply:
I tried to extract this logic and turned out that `chunk_by_size` retained no logic, just delegating methods with worse ergonomics than the caller simply calling `#slice_before` with the same logic as here.
I tried that, to move `#slice_before` to the caller, but then it became a little bit too complex at the caller site, so I chose to keep this specific implementation here, in order to encapsulate this logic.

When we have further use cases I think we'll be able to know how to extract it, as making it truly generic today didn't work quite as well.

Reply:
Okay, I think that's fair. We can always mature this a bit as we see new cases appear.